### PR TITLE
Ensure parallel jobs get assigned in one round

### PR DESCRIPTION
**Pull request by mudler**
This PR is addressing the race condition in _prefer_parallel.
~~To guarantee the cluster execution in one scheduling round it is also introducing a new variable ```PARALLEL_CLUSTER``` - when this is defined in the job settings, it will ensure the job to be assigned ONLY if the child/parents are going in the same scheduling round.~~ will be default

This has been already tested ( and still deployed ) on staging: http://e122.suse.de/tests

See related progress issue: https://progress.opensuse.org/issues/25892

Created https://progress.opensuse.org/issues/32725 for refactoring, since now that area calls for it

### CODE REVIEW

**Comment by coolo** [comment]
I still think this should be the default - I can see 0 use of a cluster not scheduled together. Can you see one?

**Comment by mudler** [comment]
yeah i can - tests on staging reveal that you can prevent cluster jobs to go running easily if you have a lot of small ones running and the workers are sharing the same worker class. Since this depends on usage, and test writing, i think this feature should be disabled when necessary

**Comment by coolo** [comment]
you have to explain this a little more I'm afraid. 'prevent cluster jobs to go running' is exactly what I want - I don't want workers to take half the cluster. Never.

**Comment by mudler** [comment]
I'll just remove the option, and we will see then :)

**Comment by coolo** [comment]
in reality we would add that variable to all PARALLEL_WITH anyway

**Comment by mudler** [comment]
But it's not present in all tests: http://e122.suse.de/tests/12350#settings

**Comment by coolo** [comment]
well, PARALLEL_WITH is only in the 'control' test - but scheduling all clients is never useful. I again ask you: what use case of a cluster test scheduled on its own do you see?

I guess the underlying problem is that the jobs are still picked from DB one by one?

**Comment by mudler** [comment]
It is not a use case, call it safety. it is e.g. better suitable for other environments, that are not production, or even if you are sharing the same machine with two different worker_class, to avoid cluster's starvation it's safer imo to disable it - but again, will just drop it and let's see.

With regrd to the query: it is still querying - but in a bit smartish way then before. I do believe that we should remove the scheduling via DB at some point, it's not a problem of running multiple queries, but instead we should run none at all.

**Comment by foursixnine** [comment]
@coolo Bingo. Problem i also see, is that sometimes a job can come and it might not have a PARALLEL_WITH, but only the parent (or control job).

The approach tries to solve that case. (You can see it with HPC tests that have two control jobs):

`PARALLEL_WITH=hpc_ganglia_supportserver`
`PARALLEL_WITH=hpc_ganglia_supportserver,hpc_ganglia_server`

Which can lead to the scheduler getting a bit confused.

**Comment by foursixnine** [comment]
Regarding the querying... we can always create a view in the database and query that instead of relying in the orm.

**Comment by coolo** [comment]
PARALLEL_CLUSTER is binary - so it doesn't help to unconfuse it.

**Comment by coolo** [comment]
@mudler I still don't follow - what you describe as 'avoid starvation' is to me 'fuckup scheduling'. Because you would schedule half a cluster then - which is pointless and broken.

**Comment by mudler** [comment]
alright then :)

**Comment by mudler** [comment]
Want to do some further test before going for it

File: lib/OpenQA/Scheduler.pm [hunk 1]
```diff
@@ -28,7 +28,7 @@ use OpenQA::Setup;
 use OpenQA::Utils 'log_debug';
 
 # How many jobs to allocate in one tick. Defaults to 50 ( set it to 0 for as much as possible)
-use constant MAX_JOB_ALLOCATION => $ENV{OPENQA_SCHEDULER_MAX_JOB_ALLOCATION} // 50;
+use constant MAX_JOB_ALLOCATION => $ENV{OPENQA_SCHEDULER_MAX_JOB_ALLOCATION} // 80;
```
Comment (ldevulder) [inline lib/OpenQA/Scheduler.pm]: Just one small thing: default has been changed to 80, but comment still says 50 :)

**Pull request merged**